### ISSUE TITLE
BatchNormalization layer runs on CPU but fails on CUDA

### ISSUE THREAD

**Issue by wmeddie** [deeplearning4j/deeplearning4j#2826]
I have a network (Full code and sample dataset here: https://github.com/wmeddie/lstm-recommender) that seems to run fine on the CPU but throws the following error when I run with nd4j-cuda-8.0.  Removing the BatchNormalization fixes the issue.

```
[info] 13:10:36.071 [main] INFO  com.yumusoft.lstmrecommender.Train$ - Starting epoch 0 of 100
[info] 13:10:36.076 [main] DEBUG o.n.j.c.CudaAffinityManager - Manually mapping thread [67] to device [0], out of [2] devices...
[info] 13:10:36.106 [main] INFO  org.nd4j.nativeblas.Nd4jBlas - Number of threads used for BLAS: 0
[error] Exception in thread "main" java.lang.IllegalArgumentException: Invalid size: cannot get size of dimension 2 for rank 2 NDArray (array shape: [40, 10])
[error]         at org.nd4j.linalg.api.ndarray.BaseNDArray.size(BaseNDArray.java:4183)
[error]         at org.deeplearning4j.nn.layers.normalization.CudnnBatchNormalizationHelper.preOutput(CudnnBatchNormalizationHelper.
java:213)
[error]         at org.deeplearning4j.nn.layers.normalization.BatchNormalization.preOutput(BatchNormalization.java:301)
[error]         at org.deeplearning4j.nn.layers.normalization.BatchNormalization.activate(BatchNormalization.java:228)
[error]         at org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doForward(LayerVertex.java:92)
[error]         at org.deeplearning4j.nn.graph.ComputationGraph.feedForward(ComputationGraph.java:1069)
[error]         at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:963)
[error]         at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:151)
[error]         at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:54)
[error]         at org.deeplearning4j.optimize.Solver.optimize(Solver.java:51)
[error]         at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:781)
[error]         at com.yumusoft.lstmrecommender.Train$$anonfun$train$1.apply$mcVI$sp(Train.scala:190)
[error]         at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:160)
[error]         at com.yumusoft.lstmrecommender.Train$.train(Train.scala:180)
[error]         at com.yumusoft.lstmrecommender.Train$.main(Train.scala:140)
[error]         at com.yumusoft.lstmrecommender.Train.main(Train.scala)
```

**Comment by raver119** [comment]
That should be something else.
>Exception in thread "main" java.lang.IllegalArgumentException: Invalid size: cannot get size of dimension 2 for rank 2 NDArray (array shape: [40, 10])

This exception says that you have 2d array of [40,10], and normalizer expects 3D at very least, since it's asking for dimension 2.

This validation was added somewhere around 0.7.2.

**Comment by raver119** [comment]
Can you please try without cudnn enabled? Just to see what happens?

**Comment by wmeddie** [comment]
Yes it runs fine after removing deeplearning4j-cuda from build.sbt.

**Comment by raver119** [comment]
@saudet what's the diff between our & cuda impl for batchnorm?

**Comment by saudet** [comment]
AFAIU, cuDNN's batch norm works only with rank 4: https://github.com/bytedeco/javacpp-presets/blob/master/cuda/src/main/java/org/bytedeco/javacpp/cudnn.java#L1628

**Comment by raver119** [comment]
Won't fix then. Workaround, obviously: stick to nd4j cuda instead of cudnn.
